Resteasy Reactive: Fix support of media types with suffixes

Sgitario opened the pull request:
Given a media type with suffix, for example: "application/test+json".

Before these changes, Resteasy Reactive was trying to find the writer/reader using only the suffix. For example, with a media type "application/test+json", it would only use "json" to locate the right writer/reader. 

Spite of this has been working well so far, if users provide a custom writer/reader for a concrete media type with suffix:

```java
@Provider
    @Produces("text/test+suffix")
    public static class SuffixMessageBodyWriter implements ServerMessageBodyWriter<Object> {

        @Override
        public boolean isWriteable(Class<?> type, Type genericType, ResteasyReactiveResourceInfo target, MediaType mediaType) {
           // ...
        }

        @Override
        public void writeResponse(Object o, Type genericType, ServerRequestContext context) {
            // ...
        }

        // ...
    }
```

The above writer will never be used. 

After these changes, we will use the media type with suffix, plus the split parts of the suffix. For example, if the media type with suffix is: "application/test+json", it will try to find the best writer/reader for "application/test+json", "application/test" and then "application/json" (In this order).

Fix https://github.com/quarkusio/quarkus/issues/15982

Sgitario commented:
cc @markusdlugi @geoand

geoand commented on independent-projects/resteasy-reactive/common/runtime/src/main/java/org/jboss/resteasy/reactive/common/core/Serialisers.java:
```diff
@@ -76,18 +76,24 @@ public List<MessageBodyReader<?>> findReaders(ConfigurationImpl configuration, C
         } while (klass != null);
 
-        return ret;
+        List<MessageBodyReader<?>> matchingReaders = new ArrayList<>();
```
Can we specify the size here?

Sgitario replied:
Done

geoand commented on independent-projects/resteasy-reactive/common/runtime/src/main/java/org/jboss/resteasy/reactive/common/core/Serialisers.java:
```diff
@@ -76,18 +76,24 @@ public List<MessageBodyReader<?>> findReaders(ConfigurationImpl configuration, C
 
-        return ret;
+        List<MessageBodyReader<?>> matchingReaders = new ArrayList<>();
+        for (ResourceReader mediaTypeMatchingReader : ret) {
```
Can we avoid the use of the iterator here? We do that in various parts of the code in order to cut down on the number of objects created while serving a request

Sgitario replied:
Done. To be precise, I reverted this change.

geoand commented on independent-projects/resteasy-reactive/common/runtime/src/main/java/org/jboss/resteasy/reactive/common/util/MediaTypeHelper.java:
```diff
@@ -267,20 +277,74 @@ public static boolean isUnsupportedWildcardSubtype(MediaType mediaType) {
         }
-        return mediaType;
+
+        List<MediaType> effectiveMediaTypes = new ArrayList<>();
```
Can we specify a size here?

Sgitario replied:
Done

geoand commented on independent-projects/resteasy-reactive/common/runtime/src/main/java/org/jboss/resteasy/reactive/common/util/MediaTypeHelper.java:
```diff
@@ -267,20 +277,74 @@ public static boolean isUnsupportedWildcardSubtype(MediaType mediaType) {
+
+    private static int countMatchingMediaTypes(List<MediaType> produces, List<MediaType> mediaTypes) {
+        int count = 0;
+        for (MediaType mediaType : mediaTypes) {
```
Same comment about the iterator here

Sgitario replied:
Fixed

geoand reviewed (commented):
Great! I added a couple small comments

Sgitario committed:
Resteasy Reactive: Fix support media types with suffixes

Given a media type with suffix, for example: "application/test+json".

Before these changes, Resteasy Reactive was trying to find the writer/reader using only the suffix. For example, with a media type "application/test+json", it would only use "json" to locate the right writer/reader. 

Spite of this has been working well so far, if users provide a custom writer/reader for a concrete media type with suffix:

```java
@Provider
    @Produces("text/test+suffix")
    public static class SuffixMessageBodyWriter implements ServerMessageBodyWriter<Object> {

        @Override
        public boolean isWriteable(Class<?> type, Type genericType, ResteasyReactiveResourceInfo target, MediaType mediaType) {
           // ...
        }

        @Override
        public void writeResponse(Object o, Type genericType, ServerRequestContext context) {
            // ...
        }

        // ...
    }
```

The above writer will never be used. 

After these changes, we will use the media type with suffix, plus the split parts of the suffix. For example, if the media type with suffix is: "application/test+json", it will try to find the best writer/reader for "application/test+json", "application/test" and then "application/json" (In this order).

Fix https://github.com/quarkusio/quarkus/issues/15982

Sgitario commented:
@geoand I think I've addressed all your comments. Thanks!

geoand reviewed (approved):
Great, thanks!

markusdlugi commented:
Thanks a lot for the fix @Sgitario, much appreciated!

geoand commented:
Good point
